Bluetooth: ISO: Reject disconnecting pending CIS as peripheral

Thalley opened the pull request:
As per the core spec, the CIS is not allowed to disconnect a CIS if it is pending (i.e. in the connecting state).

Partially fixes https://github.com/zephyrproject-rtos/zephyr/issues/61603

Thalley committed:
Bluetooth: ISO: Reject disconnecting pending CIS as peripheral

As per the core spec, the CIS is not allowed to disconnect
a CIS if it is pending (i.e. in the connecting state).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra reviewed (dismissed):
LGTM

rugeGerritsen reviewed (approved):
Looks good. Maybe the commit message should be updated to indicate that it only fixes a part of 61603?

rugeGerritsen commented:
I should have added the part that is not handled in this PR:
> The host does not handle the case that two events are raised when the central disconnects:

Thalley commented:
> I should have added the part that is not handled in this PR:
> 
> > The host does not handle the case that two events are raised when the central disconnects:

Hmm, true. I guess I didn't fully understand that part. 

Is the issue that (on the central) the CIS Established event will trigger `bt_iso_disconnected(iso);` in `hci_le_cis_established` and then when the disconnection event in `hci_disconn_complete` will trigger `LOG_ERR("Unable to look up conn with handle %u", handle);` because the CIS has already been disconnected?

rugeGerritsen commented:
> > I should have added the part that is not handled in this PR:
> > > The host does not handle the case that two events are raised when the central disconnects:
> 
> Hmm, true. I guess I didn't fully understand that part.
> 
> Is the issue that (on the central) the CIS Established event will trigger `bt_iso_disconnected(iso);` in `hci_le_cis_established` and then when the disconnection event in `hci_disconn_complete` will trigger `LOG_ERR("Unable to look up conn with handle %u", handle);` because the CIS has already been disconnected?

Yes, that is the issue I was thinking about.

Thalley commented:
> > > I should have added the part that is not handled in this PR:
> > > > The host does not handle the case that two events are raised when the central disconnects:
> > 
> > 
> > Hmm, true. I guess I didn't fully understand that part.
> > Is the issue that (on the central) the CIS Established event will trigger `bt_iso_disconnected(iso);` in `hci_le_cis_established` and then when the disconnection event in `hci_disconn_complete` will trigger `LOG_ERR("Unable to look up conn with handle %u", handle);` because the CIS has already been disconnected?
> 
> Yes, that is the issue I was thinking about.

Hmm, looking through the code again, the `bt_conn` that is associated with the `bt_iso_chan` isn't disconnected until the `hci_disconn_complete` function has been called (i.e. the `bt_conn` does not enter the `BT_CONN_DISCONNECTED` state until the event. 

Is this an issue you've seen? If so, can you elaborate on how to reproduce it (preferably in the issue)? Just by looking through the code, it doesn't seem to be an issue, but I may be missing something

rugeGerritsen commented:
I found this by observation. There are two cases:
1. CIS fails to be established, no disconnection is issues. See figure 6.46
   * `hci_le_cis_established` gets called with status != 0, which ends up with calling `bt_iso_disconnected(iso);`. 
   * `bt_iso_chan_disconnected()`, state is set to `BT_ISO_STATE_DISCONNECTED` and `chan->ops->disconnected(chan, reason);` is called.
2. CIS is disconnected from central before it has been established. 6.50 and 6.51
   * `hci_le_cis_established` gets called with status != 0, which ends up with calling `bt_iso_disconnected(iso);`. 
   * `bt_iso_chan_disconnected()`, state is set to `BT_ISO_STATE_DISCONNECTED` and `chan->ops->disconnected(chan, reason);` 
   * `hci_disconn_complete` is called. The handle will be found as it is not removed.
   * `bt_iso_disconnected(iso) is called again`

Do we have a unit test that can be modified to trigger this?

Thalley commented:
> I found this by observation. There are two cases:
> 
>     1. CIS fails to be established, no disconnection is issues. See figure 6.46
>        
>        * `hci_le_cis_established` gets called with status != 0, which ends up with calling `bt_iso_disconnected(iso);`.
>        * `bt_iso_chan_disconnected()`, state is set to `BT_ISO_STATE_DISCONNECTED` and `chan->ops->disconnected(chan, reason);` is called.
> 
>     2. CIS is disconnected from central before it has been established. 6.50 and 6.51
>        
>        * `hci_le_cis_established` gets called with status != 0, which ends up with calling `bt_iso_disconnected(iso);`.
>        * `bt_iso_chan_disconnected()`, state is set to `BT_ISO_STATE_DISCONNECTED` and `chan->ops->disconnected(chan, reason);`
>        * `hci_disconn_complete` is called. The handle will be found as it is not removed.
>        * `bt_iso_disconnected(iso) is called again`
> 
> 
> Do we have a unit test that can be modified to trigger this?

Thanks, now I understand the issue. Doesn't the call to `bt_iso_disconnected` end up triggering 
```c
	chan = iso_chan(iso);
	if (chan == NULL) {
		LOG_ERR("Could not lookup chan from disconnected ISO");
		return;
	}
```

since it the `bt_iso_chan` has already removed the association to the `bt_conn`? 

> Do we have a unit test that can be modified to trigger this?

We don't. Might be easier to add to the BSIM tests added by https://github.com/zephyrproject-rtos/zephyr/pull/61690 once that has been merged (should be fairly easy to trigger this).

rugeGerritsen commented:
> Thanks, now I understand the issue. Doesn't the call to bt_iso_disconnected end up triggering

	chan = iso_chan(iso);
	if (chan == NULL) {
		LOG_ERR("Could not lookup chan from disconnected ISO");
		return;
	}

You are right. Then the only thing remaining is that we print out an error message

Thalley committed:
Bluetooth: ISO: Handle central disc of pending CIS

If the central disconnects a CIS while it is being
established, then we receive both a CIS established event
with BT_HCI_ERR_OP_CANCELLED_BY_HOST and a disconnect complete
event.

In this case we should not call bt_iso_disconnected in the
CIS established event handler, as that will also be
called from the disconnect complete event handler.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@rugeGerritsen I modified when `hci_le_cis_established` calls `bt_iso_disconnected`, so that in the case that the central cancels the CIS connect request, we wait for the Disconnect Complete event before calling `bt_iso_disconnected`

kruithofa reviewed (commented):
LGTM
